### PR TITLE
[DROOLS-6892] Wrapper class unwrapped value is not indexed in exec-model

### PR DESCRIPTION
**Pull request by tkobayas**
**Ports** 
This PR is for 7.x
Will do for main

**JIRA**: 
https://issues.redhat.com/browse/DROOLS-6892

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>

* for a <b>full downstream build</b> 
  * for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  * for <b>github actions</b> job: add the label `run_fdb`

* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

**Comment by tkobayas** [comment]
GHA windows-latest - Java 8 : Not related to this PR
```
2022-04-12T14:06:59.4672384Z [ERROR] Tests run: 12, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 5.331 s <<< FAILURE! - in org.drools.compiler.integrationtests.TimerAndCalendarWithRealTimeTest
2022-04-12T14:06:59.4767789Z [ERROR] testIntervalRuleInsertion[KieBase type=STREAM_IDENTITY](org.drools.compiler.integrationtests.TimerAndCalendarWithRealTimeTest)  Time elapsed: 1.533 s  <<< FAILURE!
```

Linux - Compile Downstream Build : Not related to this PR
```
[2022-04-12T10:59:48.613Z] INFO: WFLYSRV0049: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) starting
[2022-04-12T10:59:48.868Z] Apr 12, 2022 6:59:48 AM org.jboss.as.server.ServerService boot
[2022-04-12T10:59:48.868Z] ERROR: WFLYSRV0055: Caught exception during boot
[2022-04-12T10:59:48.868Z] org.jboss.as.controller.persistence.ConfigurationPersistenceException: WFLYCTL0082: Failed to create backup copies of configuration file /home/jenkins/workspace/KIE/7.x/compile/drools-7.x.compile/bc/kiegroup_kie_wb_distributions/business-central-parent/business-central-webapp/target/bootable-jar-build-artifacts/wildfly/standalone/configuration/standalone.xml
[2022-04-12T10:59:48.868Z]      at org.jboss.as.controller@15.0.1.Final//org.jboss.as.controller.persistence.ConfigurationFile.successfulBoot(ConfigurationFile.java:545)
[2022-04-12T10:59:48.868Z]      at org.jboss.as.controller@15.0.1.Final//org.jboss.as.controller.persistence.BackupXmlConfigurationPersister.successfulBoot(BackupXmlConfigurationPersister.java:95)
[2022-04-12T10:59:48.868Z]      at org.jboss.as.controller@15.0.1.Final//org.jboss.as.controller.AbstractControllerService.finishBoot(AbstractControllerService.java:580)
[2022-04-12T10:59:48.868Z]      at org.jboss.as.server@15.0.1.Final//org.jboss.as.server.ServerService.finishBoot(ServerService.java:445)
[2022-04-12T10:59:48.868Z]      at org.jboss.as.server@15.0.1.Final//org.jboss.as.server.ServerService.boot(ServerService.java:415)
[2022-04-12T10:59:48.868Z]      at org.jboss.as.controller@15.0.1.Final//org.jboss.as.controller.AbstractControllerService$1.run(AbstractControllerService.java:416)
[2022-04-12T10:59:48.868Z]      at java.base/java.lang.Thread.run(Thread.java:829)
[2022-04-12T10:59:48.868Z] Caused by: java.nio.file.NoSuchFileException: /home/jenkins/workspace/KIE/7.x/compile/drools-7.x.compile/bc/kiegroup_kie_wb_distributions/business-central-parent/business-central-webapp/target/bootable-jar-build-artifacts/wildfly/standalone/configuration/standalone.xml
[2022-04-12T10:59:48.868Z]      at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
[2022-04-12T10:59:48.868Z]      at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
[2022-04-12T10:59:48.868Z]      at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
[2022-04-12T10:59:48.868Z]      at java.base/sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:548)
[2022-04-12T10:59:48.868Z]      at java.base/sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:258)
[2022-04-12T10:59:48.868Z]      at java.base/java.nio.file.Files.copy(Files.java:1295)
[2022-04-12T10:59:48.868Z]      at org.jboss.as.controller@15.0.1.Final//org.jboss.as.controller.persistence.FilePersistenceUtils.copyFile(FilePersistenceUtils.java:104)
[2022-04-12T10:59:48.868Z]      at org.jboss.as.controller@15.0.1.Final//org.jboss.as.controller.persistence.ConfigurationFile.successfulBoot(ConfigurationFile.java:539)
[2022-04-12T10:59:48.868Z]      ... 6 more
[2022-04-12T10:59:48.868Z]
[2022-04-12T10:59:48.868Z] Apr 12, 2022 6:59:48 AM org.jboss.as.server.ServerService$4 logExit
[2022-04-12T10:59:48.868Z] FATAL: WFLYSRV0056: Server boot has failed in an unrecoverable manner; exiting. See previous messages for details.
[2022-04-12T10:59:48.868Z] Apr 12, 2022 6:59:48 AM org.jboss.as.server.ApplicationServerService stop
[2022-04-12T10:59:48.868Z] INFO: WFLYSRV0050: WildFly Full 23.0.2.Final (WildFly Core 15.0.1.Final) stopped in 8ms
[2022-04-12T10:59:48.868Z] [ERROR] Error executing CLI script null
[2022-04-12T10:59:48.868Z] [ERROR] [disconnected /] embed-server --jboss-home=/home/jenkins/workspace/KIE/7.x/compile/drools-7.x.compile/bc/kiegroup_kie_wb_distributions/business-central-parent/business-central-webapp/target/bootable-jar-build-artifacts/wildfly --std-out=discard
[2022-04-12T10:59:48.868Z]
[2022-04-12T10:59:48.868Z] [INFO] FAILURE build of project org.kie:business-central-webapp
```

**Comment by tkobayas** [comment]
@mariofusco  I'm a little wondering if this issue is worth fixing. The use case is only indexing Primitive Wrapper value. Without the fix, non-exec-model can index while exec-model cannot.
```
rule R1 when
  Integer( intValue == 10 )
then
...
```
Inserting such a Primitive Wrapper class is rare in real use cases (while we may do it in unit tests). Do you think it's good to fix?

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/util/ClassUtil.java [hunk 1]
```diff
@@ -38,10 +59,35 @@ public static boolean isAccessibleProperties( Class<?> clazz, String prop ) {
...
+    }
+
+    // Used for DomainClassMetadata and index. Primitive wrapper class' immutable property (e.g. "intValue") can be indexed
+    public static List<String> getAccessiblePropertiesIncludingPrimitiveWrapper( Class<?> clazz ) {
+        if (isPrimitiveWrapper(clazz)) {
+            return PRIMITIVE_WRAPPER_PROPS_CACHE.get(clazz);
+        } else {
+            return getAccessibleProperties(clazz);
+        }
+    }
```
Comment (tkobayas) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/util/ClassUtil.java:78]: I just introduced `getAccessiblePropertiesIncludingPrimitiveWrapper` for DomainClassMetadata and index.

If we call `getAccessiblePropertiesIncludingPrimitiveWrapper` for checking settableProperties for property reactivity, properties like `intValue` would become subject to reactivity so it's not expected.

**Review by mariofusco — CHANGES_REQUESTED** [review]
@tkobayas I believe that you're focusing on a specific situation (the primitive wrappers), but in reality found a problem that is more general that this. To prove what I'm saying add the following method to the `Person` class

```
    public int calcAge() {
        return age;
    } 
```

and in your test case replace the 3 constraints in order to make them invoke the new method, i.e. changing them from

  `Person( age == 10 )`

to

  `Person( calcAge == 10 )`

Even in this case you will see that the plain drl version is indexing those constraints while the exec-model one isn't. I think that we should fix this more general scenario and if we do so also the primitive wrapper issue will be fixed as a special case of this.

My suggestion is to close this PR and work on a new one fixing the problem as I described it. Let me know if this is ok for you and if you need any help with the more general issue.

**Comment by tkobayas** [comment]
Thanks, @mariofusco 

> My suggestion is to close this PR and work on a new one fixing the problem as I described it. Let me know if this is ok for you and if you need any help with the more general issue.

Sure, I'll raise a new PR and let you know.

**Pull request closed without merging**